### PR TITLE
Drop Ruby < 2.3

### PR DESCRIPTION
**Pull request by pirj**
Better viewed ignoring indentation changes https://github.com/rspec/rspec-support/pull/435/files?diff=split&w=1

As per https://github.com/rspec/rspec/issues/61#issuecomment-687290045

Minimal changes to `rspec-core` and `rspec-mocks` were required to run `rspec-support`'s own spec suite, pushed to branches with the same name.

## What has been done

Done:
- [x] reduce Travis matrix to rubies with complete kwargs support (and compatibility issues with modern openssl)
- [x] reduce Appveyor matrix
- [x] clean up RubyFeatures
- [x] check respond_to?
- [x] check method_defined?
- [x] check send (where we call private methods that later became public)
- [x] cleanup Gemfile
- [x] update required_ruby_version
- [x] update VERSION
- [x] check RUBY_VERSION
- [x] check if we require_rspec_support ruby_features where RubyFeatures are not used
- [-] check define_method where def would suffice
- [+] check (class << self; self; end) usages
- [x] check if Java::JavaLang::String.instance_method(:char_at).arity == -1
- [?] check Java::JavaLang::String.instance_method(:char_at).parameters == [] - still the case
- [x] check 1.8, 1.9, 2.0, 2.1, jruby, 9.0, 9.1, mri
- [x] change old ruby induced rubocop settings, fix offences
- [x] check requires for support's mutex

Discuss:
- [?] introduce https://github.com/ruby/ruby2_keywords dependency and remove respond_to?(:ruby2_keywords, true) check
- [?] cleanup of build functions/predicates
- [?] drop Ruby 2.3 - problematic'ish due to a dependency on outdated OpenSSL, EOL'ed, not supported by RuboCop, still used by 8% of users

TODO:
- [ ] check "work around", "workaround" - there are two cases: check if this was fixed in JRuby 9.2
- [ ] check docs

for rspec-dev:

- [ ] remove unused predicate_functions
- [ ] remove legacy .rubocop settings

### CODE REVIEW

File: lib/rspec/support/comparable_version.rb [hunk 1]
```diff
@@ -1,46 +1 @@
...
-            segment.to_i
-          else
-            segment
-          end
-        end
-      end
-    end
-  end
-end
+# FIXME: if you remove this, `double`, `hide_const`, `stub_const` can't be found in specs
```
Comment (pirj) [inline lib/rspec/support/comparable_version.rb:1]: 🤯 🤷‍♂️

**Comment by pirj** [comment]
I have no idea what's up with Travis, the build shows as "queued", but I don't see any other builds running, including Core and Mocks' `rc4.0` branches.

**Comment by JonRowe** [comment]
Any PRs working up proposals for RSpec 4.0 need to be marked as draft as they will require very careful release management, they will not be merged to main for a long time.

**Comment by JonRowe** [comment]
I've created a `4-0-dev` branch on all the repos, please target any PRs to that branch

**Comment by pirj** [comment]
@JonRowe Sounds good, thanks!

May I kindly ask you to grant me `rspec/rspec` repo branch push permission? I can push to any other repo's branch, but not `rspec/rspec`.
Not being able to push a branch with matching name results in build errors:
```
Fetching https://github.com/rspec/rspec.git

fatal: Needed a single revision

Git error: command `git rev-parse --verify rc4.0` in directory
```

**Comment by pirj** [comment]
I'll close this PR and will open another one with a more descriptive branch name (`drop-old-rubies` or alike).

**Comment by JonRowe** [comment]
@pirj just make sure to target `4-0-dev`, I'll get it setup properly at the weekend. RSpec 3.x will continue to support the current Ruby matrix

**Comment by pirj** [comment]
Superseded by #436

**Pull request closed without merging**